8315655: [macos] Right click and dragging over a component with a popup menu will open the popup

alisenchung opened the pull request:
Issue is a mouse drag will trigger a popup in macos, but not in linux or windows.
The solution is to add a check for a mouse pressed event to show popup and prevent mouse entered events from triggering the popup

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315655](https://bugs.openjdk.org/browse/JDK-8315655): [macos] Right click and dragging over a component with a popup menu will open the popup (**Bug** - P3)


### Reviewers
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Committer) ⚠️ Review applies to [452dcac9](https://git.openjdk.org/jdk/pull/19569/files/452dcac9b742ac32cb1224d01877ed08d8184700)
 * [Prasanta Sadhukhan](https://openjdk.org/census#psadhukhan) (@prsadhuk - **Reviewer**) ⚠️ Review applies to [06a7c212](https://git.openjdk.org/jdk/pull/19569/files/06a7c2126465496022945ac627866288f70337de)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19569/head:pull/19569` \
`$ git checkout pull/19569`

Update a local copy of the PR: \
`$ git checkout pull/19569` \
`$ git pull https://git.openjdk.org/jdk.git pull/19569/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19569`

View PR using the GUI difftool: \
`$ git pr show -t 19569`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19569.diff">https://git.openjdk.org/jdk/pull/19569.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19569#issuecomment-2151149945)

honkar-jdk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,84 @@
+
+                2) Move the mouse to the JLabel with the text "to here"
+
+                3) Observe that the popup menu assigned to the inner JPanel appears
+                (macOS) or does not appear (Windows, Linux)
+                """;
```
The instruction is not clear as to what is expected to pass or fail the test. rephrase?

honkar-jdk reviewed (commented):
Fix works as expected and similar to native apps.

prsadhuk commented:
I guess it needs to be understood why it is not failing in windows and linux and only reproducible in macos. The fix in BasicLookAndFeel is not appropriate in my opinion as that shared code will be exercised in windows and linux too..

I guess we need to find out why `BasicLookAndFeel.evenDIspatched` is called for MOUSE_PRESSED in macos and not in windows/linux (ie only called for MOUSE_RELEASED event), maybe something needs to be done in macos JDK native event processing to ensure it is only called for "mouse release" event for showing popup.

Also, the test can be automated as it is about popup being shown or not which can be easily checked by `isPopupMenuVIsible/isVisible` method

alisenchung commented:
> I guess it needs to be understood why it is not failing in windows and linux and only reproducible in macos. The fix in BasicLookAndFeel is not appropriate in my opinion as that shared code will be exercised in windows and linux too..
> 
> I guess we need to find out why `BasicLookAndFeel.evenDIspatched` is called for MOUSE_PRESSED in macos and not in windows/linux (ie only called for MOUSE_RELEASED event), maybe something needs to be done in macos JDK native event processing to ensure it is only called for "mouse release" event for showing popup.
> 
> Also, the test can be automated as it is about popup being shown or not which can be easily checked by `isPopupMenuVIsible/isVisible` method

I guess the root of the issue is that MOUSE_ENTERED is a popup trigger in macos only. I can look into NSEvent isPopupTrigger and change the code to require the event to be MOUSE_PRESSED/MOUSE_RELEASED

alisenchung commented:
@prsadhuk I've updated the fix to only affect macOS code. I changed NSEvent isPopupTrigger to check for MOUSE_PRESSED or MOUSE_RELEASED java events. Please take a look

DamonGuy commented on src/java.desktop/macosx/classes/sun/lwawt/macosx/NSEvent.java:
```diff
@@ -269,7 +269,12 @@ static int nsToJavaEventType(int nsEventType) {
     static native char nsToJavaChar(char nsChar, int modifierFlags, boolean spaceKeyTyped);
 
-    static boolean isPopupTrigger(int jmodifiers) {
+    static boolean isPopupTrigger(int jmodifiers, int jeventType) {
+        if (jeventType != MouseEvent.MOUSE_PRESSED
+                && jeventType != MouseEvent.MOUSE_RELEASED) {
+            return false;
+        }
```
Just to double check, are the only jeventTypes to trigger popups using mouse? Could there be other event types, maybe using keyboard for example with tab/arrows & enter?

alisenchung replied:
I believe only MouseEvents can be popup triggers. I checked in MouseEvent that there is a field for popup triggers and no such field in KeyEvent. 

The isPopupTrigger method here also previously checked button down masks already, so adding a check for the correct mouse events shouldn't be an issue even if there was a way to show a popup menu with a different event type.

prsadhuk replied:
Although there's one more issue it seems where in native, after popup is shown when right mouse button (or CTRL+left mouse button) is pressed, if you release the mouse the popup vanishes which I noticed in Notes, Mail but it does not happen in JDK and popup remains open...but it seems not related to this...did you notice?

alisenchung replied:
Hmm, not sure why. I doubt it's related to this since both MOUSE_PRESSED and MOUSE_RELEASED are still popuptriggers before and after the fix. I can make a new issue for this to match native behavior.

alisenchung replied:
I've tested that interaction before and after the fix and it's the same behavior. The initial MOUSE_PRESSED will open a popup, then the subsequent MOUSE_RELEASED will do nothing. When the popup is already open, the MOUSE_PRESSED will close the popup, but the MOUSE_RELEASED will do nothing. So it's separate from this fix but it might be something to do with MOUSE_RELEASED or maybe MOUSE_DRAGGED/MOUSE_CLICKED cancelling the popup behavior. I can look to make a fix separate from this issue.

alisenchung commented:
I've automated the test and confirmed that it fails before the fix and passes after the fix.

DamonGuy reviewed (commented):
Don't forget to update the years for the copyrights at the top of each file changed.

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,123 @@
+    static JFrame frame;
+    static JPanel panel;
+    static JPanel innerPanel;
+    static JPopupMenu menu;
```
guess these 2 can be local vars..

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,123 @@
+/*
+ * @test
+ * @bug 8315655
+ * @requires (os.family == "mac")
```
Also, is there any need to restrict it to mac as I guess in other platforms, this test should anyway pass...Did you check?

alisenchung replied:
I've removed the restriction, the test passes on all platforms

DamonGuy commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,135 @@
+        try {
+            failed = false;
+            Robot robot = new Robot();
+            robot.setAutoDelay(100);
+
+            SwingUtilities.invokeAndWait(() -> {
+                createAndShowGUI();
+            });
+            robot.waitForIdle();
+            robot.delay(1000);
```
```suggestion
            robot.setAutoDelay(100);
            robot.setAutoWaitForIdle(true);

            SwingUtilities.invokeAndWait(() -> {
                createAndShowGUI();
            });
            robot.delay(1000);
```

I was able to remove all the individual `waitForIdle()` calls and replaced it with `setAutoWaitForIdle()`. Still seemed to work fine on macOS.

alisenchung replied:
Most tests i've looked at aren't using setAutoWaitForIdle(). Are there any issues with test stability?

DamonGuy replied:
I believe you're right. I recall mouse drag movements having some oddity with setAutoWaitForIdle. It's probably fine as is.

DamonGuy reviewed (approved):
Tested with and without the fix. Automating it does relieve the need to give a detailed pass/fail condition. Small nit suggestion.

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,135 @@
+
+    public static void main(String[] args) throws Exception {
+        try {
+            failed = false;
```
By default boolean class variable is false, no need to set explicitly..

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,135 @@
+public class MouseDragPopupTest {
+    static volatile boolean failed;
+    static volatile JFrame frame;
+    static volatile JPanel panel;
```
frame and panel are/should be accessed in EDT only so no need for volatile

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,135 @@
+ * @test
+ * @bug 8315655
+ * @key headful
+ * @run main MouseDragPopupTest
```
summary tag missing

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,135 @@
+            final Point dstPoint = new Point(srcPoint);
+            dstPoint.translate(4 * d.width / 15, 0);
+
+
```
unnecessary line

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,135 @@
+            }
+            robot.waitForIdle();
+
+            robot.mouseRelease(InputEvent.BUTTON3_DOWN_MASK);
```
Better to release in `finally` block so that if unexpectedly any exception occurs it will not hamper any other test if tun as part of CI job..

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,129 @@
+/*
+ * @test
+ * @bug 8315655
+ * @summary Right click and dragging over a component with a popup menu will open the popup
```
Actually summary should be what test is expected to do like "Verifies Right click and dragging over a component with a popup menu will not open the popup"

prsadhuk commented on test/jdk/javax/swing/JPopupMenu/MouseDragPopupTest.java:
```diff
@@ -0,0 +1,129 @@
+    static JPanel panel;
+    static Robot robot;
+    static volatile Point srcPoint;
+    static volatile Dimension d;
```
better to club all volatile vars one after another..

alisenchung commented:
/integrate
